### PR TITLE
feat: get saved card data from remote during payment confirmation

### PR DESCRIPTION
**Pull request by reidzeibel**
closes https://github.com/AtB-AS/kundevendt/issues/17817

old PR for reference : https://github.com/AtB-AS/mittatb-app/pull/4564

On this PR:
- Refactor calling payment cards API to `useQuery` and `useMutation`.
- Call API to check when getting saved card, to check if the saved card is still available on API or not.

### CODE REVIEW

File: src/ticketing/api.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export async function listRecurringPayments(): Promise<RecurringPayment[]> {
   const response = await client.get<RecurringPayment[]>(url, {
     authWithIdToken: true,
   });
-  return response.data;
+  return response.data.reverse();
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/ticketing/api.ts]: removed :D

File: src/ticketing/use-delete-recurring-payment-mutation.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import {deleteRecurringPayment} from '@atb/ticketing';
+import { LIST_RECURRING_PAYMENTS_QUERY_KEY } from './use-list-recurring-payments-query';
+
+export const DELETE_RECURRING_PAYMENT_MUTATION_KEY = 'deleteRecurringPayment';
+
+export const useDeleteRecurringPaymentMutation = () => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationKey: [DELETE_RECURRING_PAYMENT_MUTATION_KEY],
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/ticketing/use-delete-recurring-payment-mutation.tsx]: removed :D

File: src/ticketing/use-list-recurring-payments-query.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+const ONE_HOUR_MS = 1000 * 60 * 60;
+
+export const LIST_RECURRING_PAYMENTS_QUERY_KEY = 'getListRecurringPayments';
+
+export const useListRecurringPaymentsQuery = () => {
+  const {authenticationType} = useAuthState();
+
+  return useQuery({
+    queryKey: [LIST_RECURRING_PAYMENTS_QUERY_KEY],
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx]: added `abtCustomerId` to queryKey

File: src/ticketing/use-list-recurring-payments-query.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    queryFn: listRecurringPayments,
+    cacheTime: ONE_HOUR_MS,
+    enabled: authenticationType === 'phone',
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx]: removed the `enabled` part and added check before running `queryFn`

File: src/stacks-hierarchy/saved-payment-utils.ts [hunk 5]
```diff
@@ -3,17 +3,30 @@ import {SavedPaymentOption} from './types';
...
+        recurringPayments &&
+        savedMethod &&
+        savedMethod.savedType === 'recurring'
+      ) {
+        const method = recurringPayments.find(
+          (recurringPayment) =>
+            recurringPayment.id === savedMethod.recurringCard.id,
+        );
+        if (method) setPaymentMethod(savedMethod);
+      }
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/stacks-hierarchy/saved-payment-utils.ts:29]: `savedMethod` is the saved payment method that we get from storage, so this added code will check whether the saved payment method is still available on the API, before using it as the pre-selected payment method on the payment screen.

So basically: 

- get saved payment method from storage.
- compare the saved payment method to result from API/backend.
- If saved payment method still exists on the API, return it.
- If saved payment method no longer exists on the API, return nothing.

File: src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_ProfileStack/Profile_PaymentOptionsScreen.tsx [hunk 6]
```diff
@@ -139,17 +116,26 @@ export const Profile_PaymentOptionsScreen = ({
...
+                    deleteRecurringPayment(card.id)
+                  }
+                />
               </GenericSectionItem>
             ))}
           </Section>
         )}
-        {!isLoading && storedCards.length == 0 && <NoCardsInfo />}
+        {!recurringPaymentLoading &&
+          (!recurringPayment || recurringPayment?.length == 0) && (
```
[the remark this answers is not in the document]
Comment (reidzeibel) [reply src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_ProfileStack/Profile_PaymentOptionsScreen.tsx]: good catch, added an error check

File: src/stacks-hierarchy/saved-payment-utils.ts [hunk 7]
```diff
@@ -3,17 +3,32 @@ import {SavedPaymentOption} from './types';
...
+        savedMethod &&
+        savedMethod.savedType === 'recurring'
+      ) {
+        const method = recurringPayments.find(
+          (recurringPayment) =>
+            recurringPayment.id === savedMethod.recurringCard.id,
+        );
+        if (method) setPaymentMethod(savedMethod);
+      } else {
+        setPaymentMethod(savedMethod);
```
Comment (reidzeibel) [inline src/stacks-hierarchy/saved-payment-utils.ts:30]: @gorandalum I pushed the change, should now have the same functionality as before.

**Review by gorandalum — APPROVED** [review]

File: src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_ProfileStack/Profile_PaymentOptionsScreen.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (gorandalum) [reply src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_ProfileStack/Profile_PaymentOptionsScreen.tsx]: Tip for later: Rather use `isSuccess` than to check for `!isError && !isLoading`. When checking `isSuccess` you also get type safety that the returned data is not `undefined`.

File: src/ticketing/use-list-recurring-payments-query.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+export const useListRecurringPaymentsQuery = () => {
+  const {authenticationType, abtCustomerId} = useAuthState();
+
+  return useQuery({
+    queryKey: [LIST_RECURRING_PAYMENTS_QUERY_KEY, abtCustomerId, authenticationType],
+    queryFn: async () => {
+      if (authenticationType !== 'phone') {
+        return undefined
+      }
```
Comment (gorandalum) [inline src/ticketing/use-list-recurring-payments-query.tsx:17]: I just thought of something. Returning `undefined` for query function is not allowed. You haven't observed an error in Metro when returning `undefined`?

**Review by gorandalum — CHANGES_REQUESTED** [review]

File: src/ticketing/use-list-recurring-payments-query.tsx (hunk 8 above)
Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx:17]: Wait I'll double check it

Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx:17]: would this work instead? 

```
      if (authenticationType === 'phone') {
        return await listRecurringPayments();          
      }
```

So don't return anything if it is not a phone auth.

Comment (gorandalum) [reply src/ticketing/use-list-recurring-payments-query.tsx:17]: Nope, that is the same as returning `undefined`.

Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx:17]: Oh, then would ´return []´ work? 🤔

Comment (reidzeibel) [reply src/ticketing/use-list-recurring-payments-query.tsx]: Changes pushed, thanks!

**Review by gorandalum — APPROVED** [review]

**Pull request merged**